man/fhostcmd.1
.TH FHOST 1 "15 Oct 2003" FSP
.SH NAME
fhostcmd, fhost \- set environment variables
.SH SYNOPSIS
.B fhostcmd [-d delay] [-p local port] [-l local directory] [-o timeout]
[-t trace] [-w password] [-f filename] [-h number | name] [-c | -b]
[host port [directory] | abbreviation]
.LP
.B fhost [-d delay] [-p local port] [-l local directory] [-o timeout]
[-t trace] [-w password] [-f filename] [-h number | name] [-c | -b]
[host port [directory] | abbreviation]
.SH DESCRIPTION
.LP
.B fhost
uses a resources files to set all environmant variables needed to connect to a
given host via FSP. If a -f option is given, the given filename is also used
as resources file.
.B fhost
loads all available resource files. If server is defined in multiple
files, first match is returned. Load order is
.I .fsp_prof
in the current working directory and in the user's home
directory. Then a system wide default file is used.
If
.B fhost
is invoked with an abbreviation given, the resource files are searched for the
first host to match this abbreviation. The values given for this host will
be set.
.LP
If
.B fhost
is invoked with a given host and port number, this host will be used and no
searching in the resource file will be done.
.LP
Setting up the host the user can change between using the host name or the
internet address. In case only one is given and the other should be used
.B fhost
tries to find it using a built in mini-nslookup facility. The default is to use
what is given and to use the number if both are given.
.LP
Since programs cannot change environment variables of their parent shell in
a simple way, it is recommended that the user defines a shell alias or
function to make it possible.
.B csh
example:
.LP
.nf
alias fhost 'eval `fhostcmd \!*`'
.fi
.SH FSP_PROF
For more information about the resource file confer
.I fsp_prof(5).
Here's just a short example of how it works. If your resource file contains
the following

.nf
host seismo.soar.cs.cmu.edu
fsp port 30
local directory ~/sources/fsp-266.1
trace off
.fi

a call of
.B fhost
seismo in
.B csh
will produce the following output to be evaluated

.nf
unsetenv FSP_TRACE;
setenv FSP_PORT 30;
setenv FSP_LOCAL_DIR ~/sources/fsp-266.1;
setenv FSP_HOST seismo.soar.cs.cmu.edu;
setenv FSP_DIR /;
.fi
.SH ARGUMENTS
.LP
.B -d
-- set FSP_DELAY to
.I delay
.LP
.B -p
-- set FSP_LOCALPORT to
.I local port
.LP
.B -l
-- set FSP_LOCAL_DIR to
.I local directory
.LP
.B -o
-- set FSP_TIMEOUT to
.I timeout
.LP
.B -t
-- set FSP_TRACE if
.I trace
=
.I on,
unset it if
.I trace
=
.I off
.LP
.B -p
-- set FSP_PASSWORD to
.I password
.LP
.B -f
-- use
.I filename
as resource file
.LP
.B -b
-- output Bourne shell code
.LP
.B -c
-- output C shell code
.SH ENVIRONMENT
.LP
The environment variables associate this program with a FSP database and are
set (resp. unset) if needed. See fsp_env(7) for list.
.SH "SEE ALSO"
.PD
fcatcmd(1), fcdcmd(1), fgetcmd(1), fgrabcmd(1), flscmd(1), fmkdir(1),
fprocmd(1), fput(1), frmcmd(1), frmdircmd(1), fver(1), fducmd(1),
fhostcmd(1), fspd(1), fsp_prof(5), ffindcmd(1), fsp_env(7), fspsites(5)


ChangeLog
Version 2.8.1b27 - 9 Sep 2014
	fhostcmd: Fixed output of FSP_TRACE variable for Bourne style shells
	correctly initialize to zero client packet sent statistic counter
	note default timeout values in fsp_env man page
	setting environment value FSP_TIMEOUT to <=0 uses 4byte maxint value
	note default fspd max packet size in fsp_env man page
	fix build on systems with utmpx.h API (FreeBSD 9)
	fspscan: do not shadow host variable
	fspscan: fix clang warning about empty for cycle body by placing
	  semicolon on separate line
	FAQ: Cleaned lot of SGML markup errors
	Do not use sgmlformat, call jade directly
	correctly zero entire sockaddr_in structure during name resolving,
	  error found by Clang.
	allow docbook dsssl stylesheets to be specified at scons command line
	always validate path in SETPRO and MKDIR even if packet is resent and
	  no action will be taken. This is needed to avoid uninitialised
	  pointer access during reply. Found by Clang.

Version 2.8.1b26 - 1 Oct 2009
	fver now reports if server can process extra data in input packets.
	  fspclient did that for a long time. This field is used by fspd
	  2.8.1 beta 21 or newer.
	Server version string sent as response to CC_VERSION command have
	  no longer \n appended to it. Old FSP used it but fver client
	  command can handle server strings without \n for quite a long
	  time, so i decided to remove it. It is bad idea to terminate
	  version strings by \n while it is not required by FSP protocol.
	glob() function used in FSP clients renamed to bsdglob because it
	  clashed with glob() function in newer Linux glibc resulting in
          segfaults.
	non continuous upload detection fixed. Now detection works as follows:
	  if you are trying to upload past end of file, fspd will report that
	  you are trying to create hole in file.
	  if you are seeking backwards more than 1 packet size fspd will
	  report noncontinuous upload. Most likely port from old unfinished
	  download was reused.
	We can now reupload last block which can happen if fspd server is
	  restarted and reply to upload command is lost on way from server
	  to client.
	Check for overwriting uploaded files in fspd was reversed by mistake.
	  Now if you want to overwrite file you must have permission for
	  deleting and adding files or be directory owner.
	Turning on debug in SCons turns on maintainer mode also.
	Upload starting at zero position are now always allowed no matter if
	  source port was reused or not. Old code did that but not over
	  server restarts.
	INFO file updated with fresh information
	fsetupcmd and fhostcmd now reject port number > 65535
	fsetupcmd - code cleaned a bit. Error message is now
	  fsetup: No host given! instead of fhost:
	fsetupcmd got -l option. It lists all known fsp servers collected
	  from .fspsites files.
	fpwd command from setup.sh now prints site name instead of ip
	  address if possible.
	SCons now check if CC compiler works. We need SCons 1.1 or greater
	  for building FSP now.
	Large file support on Linux is now correctly detected.
	Error codes added to CC_ERR command in PROTOCOL. It was long wished
	  feature from ISV. Now there is standard way how to pass error codes
	  from server to client. No such codes are defined yet, but you can
	  use custom error codes in range 0xF000 - 0xFFFF. because no such
	  codes are defined fspd do not sends them in error message, but it
	  now correctly zeroes bb_pos field in CC_ERR reply to indicate
	  that extended error code status is not attached to message. Old
	  fspd versions left this field untouched - it will contain request
	  bb_pos in most, but not all cases. Because we indicate size of
	  extra data to be received, this extended error handling is
	  compatible with "plain FSP over serial" standard.
	Code portability improved. Now we don't need to check sizes of
	  predefined types because we are now using ISO C types. Only thing
	  what we need to know is if largefiles are supported or not.
	  If SCons cant run test program for sizing off_t for any reason,
	  it defaults to 4 (no largefiles, 32-bit system).

Version 2.8.1b25 - 10 Sep 2009
	GNU autotools build system was entirely replaced by SCons
	corrected misleading comment in client/lib.c
	default packetsize changed to do best fit into maximum 1500 MTU
	  Ethernet frame = 1460. This is needed for speedup the booting of
	  Etherboot clients
	PROTOCOL: when computing checksum declare *t as unsigned char * type
	changed typo setguid -> setgid in fspd.conf
	fspd.conf: added note that only numeric uid values works
	fspd.conf: added more information about packetsize
	fspd.1: updated information about packetsize
	fspd.conf: repaired some spelling errors by ispell
	MTU changed from 1500 to 1492 for better operation on pppOE networks
	Max payload changed from 4096 to 4380 bytes (3 packets)
	fspd: command line argument now overrides config file
	updated year in copyright
	fsetup command added to setup.sh
	fsetup command added to setup.csh
	fcd: remove trailing \r\n from readme file
	fcd: do not print empty line after directory protection
	fsetupcmd: don't crash if called with numeric address
	fspscan: committed fix by Yang Hong
	SCONS: turning on maintainer-mode links against efence (if possible)
	  same thing like ./configure do
	fix compilation on systems without large files
	cygwin support disabled. we cant detect it from SCons yet
	frmdir works now. It didn't worked due to adding new access rights
	  DIR_NOLIST and DIR_NOGET
	MACHINES updated. Builds confirmed on Ubuntu 8.04.3,
	  FreeBSD 6.4 and FreeBSD 7.2
	fsetupcmd added to fspmerge client
	SCons can now build SGML FSP FAQ if sgmlfmt is found during build
	  process
        fspmerge now prints short help text if it is invoked as fspmerge
	FILES is up to date now

Version 2.8.1b24 - 19 Feb 2005
	use getopt from stdio.h instead of getopt.h in fspscan.c
	  fspscan now compiles on AIX and other OS without glibc (hoaxter)

clients/fhostcmd.c
    /*********************************************************************\
    *  Copyright (c) 1993 by Michael Meskes                               *
    *  (meskes@ulysses.informatik.rwth-aachen.de)                         *
    *  Copyright (c) 2003,2006 by Radim Kolar (hsn@cybermail.net)         *
    *                                                                     *
    *  You may copy or modify this file in any manner you wish, provided  *
    *  that this notice is always included, and that you hold the author  *
    *  harmless for any loss or damage resulting from the installation or *
    *  use of this software.                                              *
    \*********************************************************************/

#include "tweak.h"
#include "client_def.h"
#include "c_extern.h"
#include <stdio.h>
#include "my-string.h"
#include "merge.h"
#include <pwd.h>
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#include <stdlib.h>
#include <ctype.h>

#ifdef HOST_LOOKUP
#include <netdb.h>
#endif

#define FSP_STAT stat


#include "fhost.h"

static const char *home="/";
static int tryfile=0;


/* generated lex parser */
extern FILE *yyin;
int yylex(void);
int yywrap(void);

static void host_usage (void) /* print usage message */
{
  fprintf(stderr,"Usage: fhost [-d delay] [-p local port] [-l local dir]\n");
  fprintf(stderr,"             [-o timeout] [-t trace] [-w password]\n");
  fprintf(stderr,"             [-f filename] [-h [number|name]] [-c | -b]\n");
  fprintf(stderr,"             [host port [directory] | abbreviation]\n");
  exit(EX_OK);
}

/* get data out of resource file */
static void parse_prof_file_new (const char * filename)
{

  FILE *input=NULL;
  int rc;

  if(filename)
  {
      input=fopen(filename,"r");
  }

  if(input)
  {
      yyin=input;
      rc=0;
  } else
      rc=yywrap();

  if(rc==0)
      yylex();

  if(input)
      fclose(input);
}

int main (int argc, char ** argv)
{
  int optletter,csh,lhost=0;
  register char *p;
  const char *filename=NULL;
  char *log;
  struct passwd *pw=0L;
  struct fsp_host *setup=NULL;

  log = (char *)getlogin();
  if (log) pw = getpwnam(log);
  if (!pw) pw = getpwuid(getuid());
  if (pw) {
    csh = !strcmp(pw->pw_shell + strlen(pw->pw_shell) - 3, "csh");
    home = pw->pw_dir;   /* for default search for file .fsp_prof*/
  } else
      home=getenv("HOME");

  /*
   * Figure out what shell we're using.  A hack, we look for a shell
   * ending in "csh".
   */
  log=getenv("SHELL");
  if(log)
  {
    csh = !strcmp(log + strlen(log) - 3, "csh");
  }

  setup=init_host();
  while ((optletter=getopt(argc, argv,"d:p:l:t:o:f:h:w:bc?")) != EOF)
    switch (optletter) {
      case '?':
        host_usage();
      case 'd':
        setup->delay=atol(optarg); /* FSP_DELAY */
	break;
      case 'p':
	setup->local_port=atol(optarg); /* FSP_LOCAL_PORT */
	break;
      case 'b':
	csh=0;
	break;
      case 'c':
	csh=1;
	break;
      case 'w':
	setup->password=optarg; /* FSP_PASSWORD */
	break;
      case 'l':
	setup->local_dir=optarg; /* FSP_LOCAL_DIR */
	break;
      case 'o':
	setup->timeout=atol(optarg); /* FSP_TIMEOUT */
	break;
      case 't':
	if (!strcmp(optarg,"on")) setup->trace=1;  /* FSP_TRACE */
	else if (!strcmp(optarg,"off")) setup->trace=0;
	else host_usage();
	break;
      case 'f':
	filename=optarg; /* file name */
	break;
      case 'h':
	if (!strcmp(optarg,"number")) lhost=NUMBER; /* host mode */
	else if (!strcmp(optarg,"name")) lhost=NAME;
	else host_usage();
	break;
      default:
	host_usage();
	break;
    }

  if(argc > optind+1 && !filename) { /* host and port and no filename given */
    for (p=argv[optind];!setup->hostname && *p && *p!='\n';p++)
      if (!isdigit(*p) && *p!='.') setup->hostname=argv[optind];
    if (!setup->hostname) setup->hostaddr=argv[optind];
    setup->port=atol(argv[optind+1]);
    if(setup->port==0 || setup->port > 65535)
	setup=init_host();
    if (argc > optind + 1) setup->dir=argv[optind+2]; /* directory given, too */
  } else if (argc > optind) { /* abbreviation given */
    parse_prof_file_new(filename);
    setup=find_host(argv[optind]);
    if(!setup) setup=init_host();
  } else { /* list or set command-line options */
    if (filename || argc==1) {  /* list only */
      parse_prof_file_new(filename);
      list_prof_file();
      exit(EX_OK);
    }
  }
  if(setup->hostname==NULL && setup->hostaddr==NULL)
  {
	fprintf(stderr,"fhost: No host given!\n");
	exit(EX_USAGE);
  }
  print_host_setup(setup,csh,lhost);
  exit(EX_OK);
}

/*
 *      search order: curdir, homedir, sysdir
 *
 * Returns: 1 for terminating scanner or 0 for switching
 */

int yywrap(void)
{
  char *f2=NULL;
  int rc;

  if(yyin!=NULL)
  {
      fclose(yyin);
      yyin=NULL;
  }

  switch(tryfile)
  {
      case 0:
	     /* file in cur. dir */
             yyin=fopen(FSPPROF,"r");
	     break;
      case 1:
	     /* file in home dir */
	     f2=(char *)malloc (strlen(home) + strlen(FSPPROF) + 2);
	     if (!f2) {
		perror("malloc");
		return(1);
	     }
             sprintf (f2,"%s/%s",home,FSPPROF);
	     yyin=fopen(f2,"r");
	     free(f2);
	     break;
      case 2:
	     yyin=fopen(FSPRC,"r");
	     break;
      default:
	     return 1;
  }
  tryfile++;
  if(yyin==NULL)
  {
      /* try next available */
      rc=yywrap();
      return rc;
  }

  return 0;
}
